packages/core/src/util/ng_dev_mode.ts

```diff
@@ -98,15 +112,29 @@ export function ngDevModeResetPerfCounters(): NgDevModePerfCounters {
 }
 
 /**
- * This checks to see if the `ngDevMode` has been set. If yes,
+ * This function checks to see if the `ngDevMode` has been set. If yes,
  * then we honor it, otherwise we default to dev mode with additional checks.
  *
  * The idea is that unless we are doing production build where we explicitly
  * set `ngDevMode == false` we should be helping the developer by providing
  * as much early warning and errors as possible.
  *
- * NOTE: changes to the `ngDevMode` name must be synced with `compiler-cli/src/tooling.ts`.
+ * `ɵɵdefineComponent` is guaranteed to have been called before any component template functions
+ * (and thus Ivy instructions), so a single initialization there is sufficient to ensure ngDevMode
+ * is defined for the entire instruction set.
+ *
+ * When using checking `ngDevMode` on toplevel, always init it before referencing it
+ * (e.g. `((typeof ngDevMode === 'undefined' && initNgDevMode()) || ngDevMode)`), otherwise you can
+ *  get a `ReferenceError` like in https://github.com/angular/angular/issues/31595.
+ *
+ * Details on possible values for `ngDevMode` can be found on its docstring.
+ *
+ * NOTE:
+ * - changes to the `ngDevMode` name must be synced with `compiler-cli/src/tooling.ts`.
  */
-if (typeof ngDevMode === 'undefined' || ngDevMode) {
-  ngDevModeResetPerfCounters();
+export function initNgDevMode() {
+  if (typeof ngDevMode === 'undefined' || ngDevMode) {
+    return ngDevModeResetPerfCounters();
```

Comment:
```suggestion
    ngDevModeResetPerfCounters();
    return !!ngDevMode;
```

`ngDevModeResetPerfCounters()` always returns a counter object, even if `ngDevMode` is disabled.

Reply:
Done